tsl/src/chunk.c

```diff
@@ -536,25 +703,323 @@ chunk_copy_data(PG_FUNCTION_ARGS)
+	ts_catalog_restore_user(&sec_ctx);
+}
+
+static void
+chunk_move_activity_insert(ChunkMoveData *ccd)
+{
+	Catalog *catalog;
+	Relation rel;
+
+	catalog = ts_catalog_get();
+	rel = table_open(catalog_get_table_id(catalog, CHUNK_MOVE_ACTIVITY), RowExclusiveLock);
+
+	chunk_move_activity_insert_rel(rel, ccd);
+	table_close(rel, RowExclusiveLock);
+}
+
+static ScanTupleResult
+chunk_move_activity_tuple_update(TupleInfo *ti, void *data)
+{
+	ChunkMoveData *ccd = data;
+	Datum values[Natts_chunk_move_activity];
+	bool nulls[Natts_chunk_move_activity];
+	CatalogSecurityContext sec_ctx;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple;
+
+	heap_deform_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls);
+
+	/* We only update the "stage" and time related fields */
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_stage)] =
+		CStringGetTextDatum(ccd->stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_prev_stage_end)] =
+		TimestampGetDatum(GetCurrentTimestamp());
+
+	new_tuple = heap_form_tuple(ts_scanner_get_tupledesc(ti), values, nulls);
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_update_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti), new_tuple);
+	ts_catalog_restore_user(&sec_ctx);
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+static int
+chunk_move_activity_scan_update_by_id(int32 id, tuple_found_func tuple_found, void *data,
+									  LOCKMODE lockmode)
+{
+	Catalog *catalog = ts_catalog_get();
+	ScanKeyData scankey[1];
+	ScannerCtx scanctx = {
+		.table = catalog_get_table_id(catalog, CHUNK_MOVE_ACTIVITY),
+		.index = catalog_get_index(catalog, CHUNK_MOVE_ACTIVITY, CHUNK_MOVE_ACTIVITY_PKEY_IDX),
+		.nkeys = 1,
+		.limit = 1,
+		.scankey = scankey,
+		.data = data,
+		.tuple_found = tuple_found,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+	};
+
+	ScanKeyInit(&scankey[0],
+				Anum_chunk_move_activity_pkey_idx_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(id));
+
+	return ts_scanner_scan(&scanctx);
+}
+
+int
+chunk_move_activity_update(ChunkMoveData *ccd)
+{
+	return chunk_move_activity_scan_update_by_id(ccd->id,
+												 chunk_move_activity_tuple_update,
+												 ccd,
+												 RowExclusiveLock);
+}
+
+void
+chunk_perform_distributed_move(Oid chunk_relid, bool verbose, const char *src_node,
+							   const char *dst_node, bool delete_on_src_node)
+{
+	Chunk *chunk;
+	Cache *hcache;
+	Hypertable *ht;
+	ForeignServer *src_server, *dst_server;
+	MemoryContext old, mcxt;
+	NameData operation_id;
+	ChunkMoveData ccd;
+
+	/*
+	 * The chunk and foreign server info needs to be on a memory context
+	 * that will survive moving to a new transaction for each stage
+	 */
+	mcxt = AllocSetContextCreate(PortalContext, "chunk move activity", ALLOCSET_DEFAULT_SIZES);
```

Comment:
What is the lifetime of the PortalContext and what happens in case of failures below? I guess the memory will still be allocated as long as the session remains or is it auto-freed in case of error? I worry that this leads to  a memory leak as long as the session is active.

Reply:
Since it is a child of PortalContext, it will go away eventually even if we suffer an error; there's no need for special abort cleanup logic. You can check for similar usages in `ReindexMultipleTables` in PG code as an example.